File: desktop/src/runtime/voice-activity-detector.ts
export interface VoiceActivityConfig {
  startThreshold: number
  stopThreshold: number
  attackMs: number
  releaseMs: number
}

export type VoiceActivityEvent = 'none' | 'started' | 'ended' | 'invalid'

const defaults: VoiceActivityConfig = {
  startThreshold: 0.04,
  stopThreshold: 0.02,
  attackMs: 30,
  releaseMs: 350,
}

/** Desktop reference for the same local, model-independent gate used by RuntimeCore. */
export class VoiceActivityDetector {
  private active = false
  private candidateSince?: number
  private lastObservedAt?: number

  constructor(private readonly config: VoiceActivityConfig = defaults) {
    if (!Number.isFinite(config.startThreshold)
      || !Number.isFinite(config.stopThreshold)
      || config.startThreshold <= config.stopThreshold
      || config.stopThreshold < 0
      || config.startThreshold > 1
      || config.attackMs < 0
      || config.releaseMs < 0) {
      throw new Error('invalid voice activity configuration')
    }
  }

  observe(level: number, nowMs: number): VoiceActivityEvent {
    if (!Number.isFinite(level) || level < 0 || level > 1
      || !Number.isFinite(nowMs)
      || (this.lastObservedAt !== undefined && nowMs < this.lastObservedAt)) {
      return 'invalid'
    }
    this.lastObservedAt = nowMs
    const candidate = this.active
      ? level <= this.config.stopThreshold
      : level >= this.config.startThreshold
    if (!candidate) {
      this.candidateSince = undefined
      return 'none'
    }
    this.candidateSince ??= nowMs
    const required = this.active ? this.config.releaseMs : this.config.attackMs
    if (nowMs - this.candidateSince < required) return 'none'
    this.active = !this.active
    this.candidateSince = undefined
    return this.active ? 'started' : 'ended'
  }

  reset() {
    this.active = false
    this.candidateSince = undefined
    this.lastObservedAt = undefined
  }

  get isActive() {
    return this.active
  }
}

File: unreal/GahyeonStage/Source/GahyeonRuntimeCore/Private/VoiceActivityDetector.cpp
#include "Gahyeon/VoiceActivityDetector.h"

#include <cmath>
#include <stdexcept>

namespace Gahyeon {

VoiceActivityDetector::VoiceActivityDetector(VoiceActivityConfig config)
    : config_(config) {
    if (!std::isfinite(config_.StartThreshold)
        || !std::isfinite(config_.StopThreshold)
        || config_.StartThreshold <= config_.StopThreshold
        || config_.StopThreshold < 0.0
        || config_.StartThreshold > 1.0
        || config_.AttackMs < 0
        || config_.ReleaseMs < 0) {
        throw std::invalid_argument("invalid voice activity configuration");
    }
}

VoiceActivityEvent VoiceActivityDetector::Observe(double level, Millis nowMs) {
    if (!std::isfinite(level) || level < 0.0 || level > 1.0
        || (lastObservedAt_.has_value() && nowMs < lastObservedAt_.value())) {
        return VoiceActivityEvent::Invalid;
    }
    lastObservedAt_ = nowMs;
    const bool candidate = active_
        ? level <= config_.StopThreshold
        : level >= config_.StartThreshold;
    if (!candidate) {
        candidateSince_.reset();
        return VoiceActivityEvent::None;
    }
    if (!candidateSince_.has_value()) candidateSince_ = nowMs;
    const Millis required = active_ ? config_.ReleaseMs : config_.AttackMs;
    if (nowMs - candidateSince_.value() < required) return VoiceActivityEvent::None;

    active_ = !active_;
    candidateSince_.reset();
    return active_ ? VoiceActivityEvent::Started : VoiceActivityEvent::Ended;
}

void VoiceActivityDetector::Reset() {
    active_ = false;
    candidateSince_.reset();
    lastObservedAt_.reset();
}

bool VoiceActivityDetector::Active() const {
    return active_;
}

} // namespace Gahyeon
